File: Hope/Hope/image_hashing.py
#!/usr/bin/env python3
import sys
from PIL import Image
import numpy as np
from typing import Tuple, Optional
from pathlib import Path
import json


class ImageHasher:
    def __init__(self, hash_size: int = 8):
        self.hash_size = hash_size
    
    def _prepare_image(self, image_path: str, size: Tuple[int, int]) -> Image.Image:
        img = Image.open(image_path).convert('L')  # Convert to grayscale
        img = img.resize(size, Image.Resampling.LANCZOS)
        return img
    
    def average_hash(self, image_path: str) -> str:
        img = self._prepare_image(image_path, (self.hash_size, self.hash_size))
        pixels = np.array(img, dtype=np.float32)
        
        avg = pixels.mean()
        hash_bits = (pixels > avg).flatten()
        
        return self._bits_to_hex(hash_bits)
    
    def difference_hash(self, image_path: str) -> str:
        img = self._prepare_image(image_path, (self.hash_size + 1, self.hash_size))
        pixels = np.array(img, dtype=np.float32)
        
        diff = pixels[:, 1:] > pixels[:, :-1]
        hash_bits = diff.flatten()
        
        return self._bits_to_hex(hash_bits)
    
    def perceptual_hash(self, image_path: str) -> str:
        img = self._prepare_image(image_path, (32, 32))
        pixels = np.array(img, dtype=np.float32)
        
        dct = self._dct2d(pixels)
        
        dct_low = dct[:self.hash_size, :self.hash_size]
        
        median = np.median(dct_low)
        hash_bits = (dct_low > median).flatten()
        
        return self._bits_to_hex(hash_bits)
    
    def _dct2d(self, img: np.ndarray) -> np.ndarray:
        try:
            from scipy.fftpack import dct
            return dct(dct(img.T, norm='ortho').T, norm='ortho')
        except ImportError:
            return self._simple_dct2d(img)
    
    def _simple_dct2d(self, img: np.ndarray) -> np.ndarray:
        M, N = img.shape
        dct_matrix = np.zeros((M, N))
        
        for u in range(M):
            for v in range(N):
                sum_val = 0
                for i in range(M):
                    for j in range(N):
                        sum_val += img[i, j] * \
                                   np.cos((2 * i + 1) * u * np.pi / (2 * M)) * \
                                   np.cos((2 * j + 1) * v * np.pi / (2 * N))
                
                cu = 1 / np.sqrt(M) if u == 0 else np.sqrt(2 / M)
                cv = 1 / np.sqrt(N) if v == 0 else np.sqrt(2 / N)
                dct_matrix[u, v] = cu * cv * sum_val
        
        return dct_matrix
    
    def _bits_to_hex(self, bits: np.ndarray) -> str:
        binary_str = ''.join(['1' if b else '0' for b in bits])
        
        while len(binary_str) % 4 != 0:
            binary_str += '0'
        
        hex_str = hex(int(binary_str, 2))[2:]
        
        expected_len = (len(bits) + 3) // 4
        return hex_str.zfill(expected_len)
    
    def compute_hash(self, image_path: str, method: str = "phash") -> str:
        method = method.lower()
        
        if method == "ahash":
            return self.average_hash(image_path)
        elif method == "dhash":
            return self.difference_hash(image_path)
        elif method == "phash":
            return self.perceptual_hash(image_path)
        else:
            raise ValueError(f"Unknown hash method: {method}. Use 'ahash', 'dhash', or 'phash'")
    
    def compute_all_hashes(self, image_path: str) -> dict:
        return {
            "ahash": self.average_hash(image_path),
            "dhash": self.difference_hash(image_path),
            "phash": self.perceptual_hash(image_path)
        }


def hamming_distance(hash1: str, hash2: str) -> int:
    if len(hash1) != len(hash2):
        raise ValueError("Hashes must be the same length")
    
    int1 = int(hash1, 16)
    int2 = int(hash2, 16)
    xor = int1 ^ int2
    
    distance = bin(xor).count('1')
    return distance


def compare_hashes(hash1: str, hash2: str, max_bits: int = 64) -> float:
    distance = hamming_distance(hash1, hash2)
    similarity = 1.0 - (distance / max_bits)
    return similarity


def is_similar(hash1: str, hash2: str, threshold: float = 0.9, max_bits: int = 64) -> bool:
    similarity = compare_hashes(hash1, hash2, max_bits)
    return similarity >= threshold


def save_hash_to_file(image_path: str, output_path: str, method: str = "phash"):
    hasher = ImageHasher()
    hash_value = hasher.compute_hash(image_path, method)
    
    hash_data = {
        "image_path": str(Path(image_path).absolute()),
        "method": method,
        "hash": hash_value,
        "hash_size": hasher.hash_size
    }
    
    with open(output_path, 'w') as f:
        json.dump(hash_data, f, indent=2)


def load_hash_from_file(hash_file: str) -> dict:
    with open(hash_file, 'r') as f:
        return json.load(f)


def verify_image(image_path: str, hash_file: str, threshold: float = 0.9) -> Tuple[bool, float]:
    hash_data = load_hash_from_file(hash_file)
    stored_hash = hash_data["hash"]
    method = hash_data["method"]
    hash_size = hash_data.get("hash_size", 8)
    
    hasher = ImageHasher(hash_size=hash_size)
    current_hash = hasher.compute_hash(image_path, method)
    
    max_bits = hash_size * hash_size
    similarity = compare_hashes(stored_hash, current_hash, max_bits)
    is_match = similarity >= threshold
    
    return is_match, similarity


def main():
    import argparse
    
    parser = argparse.ArgumentParser(description="Image Hashing Tool")
    parser.add_argument("--input", required=True, help="Input image path")
    parser.add_argument("--output", help="Output hash file (JSON)")
    parser.add_argument("--method", default="phash", choices=["ahash", "dhash", "phash"],
                        help="Hash method to use")
    parser.add_argument("--compare", help="Compare with another image")
    parser.add_argument("--verify", help="Verify against a hash file")
    parser.add_argument("--threshold", type=float, default=0.9, 
                        help="Similarity threshold for verification")
    
    args = parser.parse_args()
    
    hasher = ImageHasher()
    
    if args.compare:
        hash1 = hasher.compute_hash(args.input, args.method)
        hash2 = hasher.compute_hash(args.compare, args.method)
        similarity = compare_hashes(hash1, hash2)
        
        print(f"Image 1: {args.input}")
        print(f"Hash 1:  {hash1}")
        print(f"Image 2: {args.compare}")
        print(f"Hash 2:  {hash2}")
        print(f"Similarity: {similarity:.2%}")
        print(f"Hamming Distance: {hamming_distance(hash1, hash2)}")
        
    elif args.verify:
        is_match, similarity = verify_image(args.input, args.verify, args.threshold)
        
        print(f"Image: {args.input}")
        print(f"Hash file: {args.verify}")
        print(f"Similarity: {similarity:.2%}")
        print(f"Match: {'YES' if is_match else 'NO'} (threshold: {args.threshold:.2%})")
        
        sys.exit(0 if is_match else 1)
        
    else:
        if args.output:
            save_hash_to_file(args.input, args.output, args.method)
            print(f"Hash saved to: {args.output}")
        
        hash_value = hasher.compute_hash(args.input, args.method)
        print(f"Image: {args.input}")
        print(f"Method: {args.method}")
        print(f"Hash: {hash_value}")


if __name__ == "__main__":
    main()

File: Hope/Hope/ImageHashing.cs
using System;
using System.IO;
using System.Text;
using System.Windows.Media.Imaging;
using System.Text.Json;

namespace ArtShield
{
    public class ImageHashing
    {
        private readonly int hashSize;
        /// <param name="hashSize">Size of the hash grid </param>
        public ImageHashing(int hashSize = 8)
        {
            this.hashSize = hashSize;
        }
        public enum HashMethod
        {
            AverageHash,
            DifferenceHash,
            PerceptualHash
        }


        /// <param name="imagePath">Path to the img</param>
        /// <param name="method">Hash method</param>
        public string ComputeHash(string imagePath, HashMethod method = HashMethod.PerceptualHash)
        {
            return method switch
            {
                HashMethod.AverageHash => ComputeAverageHash(imagePath),
                HashMethod.DifferenceHash => ComputeDifferenceHash(imagePath),
                HashMethod.PerceptualHash => ComputePerceptualHash(imagePath),
                _ => throw new ArgumentException($"Unknown hash method: {method}")
            };
        }
        private string ComputeAverageHash(string imagePath)
        {
            byte[,] pixels = LoadAndResizeGrayscale(imagePath, hashSize, hashSize);
            double sum = 0;
            for (int i = 0; i < hashSize; i++)
            {
                for (int j = 0; j < hashSize; j++)
                {
                    sum += pixels[i, j];
                }
            }
            double average = sum / (hashSize * hashSize);
            bool[] bits = new bool[hashSize * hashSize];
            int index = 0;
            for (int i = 0; i < hashSize; i++)
            {
                for (int j = 0; j < hashSize; j++)
                {
                    bits[index++] = pixels[i, j] > average;
                }
            }

            return BitsToHex(bits);
        }
        private string ComputeDifferenceHash(string imagePath)
        {
            byte[,] pixels = LoadAndResizeGrayscale(imagePath, hashSize + 1, hashSize);
            bool[] bits = new bool[hashSize * hashSize];
            int index = 0;
            for (int i = 0; i < hashSize; i++)
            {
                for (int j = 0; j < hashSize; j++)
                {
                    bits[index++] = pixels[i, j + 1] > pixels[i, j];
                }
            }
            return BitsToHex(bits);
        }
        private string ComputePerceptualHash(string imagePath)
        {
            byte[,] pixels = LoadAndResizeGrayscale(imagePath, 32, 32);
            double[,] imageData = new double[32, 32];
            for (int i = 0; i < 32; i++)
            {
                for (int j = 0; j < 32; j++)
                {
                    imageData[i, j] = pixels[i, j];
                }
            }
            double[,] dct = ComputeDCT(imageData);
            double[] lowFreq = new double[hashSize * hashSize];
            int index = 0;
            for (int i = 0; i < hashSize; i++)
            {
                for (int j = 0; j < hashSize; j++)
                {
                    lowFreq[index++] = dct[i, j];
                }
            }
            Array.Sort(lowFreq);
            double median = lowFreq[lowFreq.Length / 2];
            bool[] bits = new bool[hashSize * hashSize];
            index = 0;
            for (int i = 0; i < hashSize; i++)
            {
                for (int j = 0; j < hashSize; j++)
                {
                    bits[index++] = dct[i, j] > median;
                }
            }

            return BitsToHex(bits);
        }
        private byte[,] LoadAndResizeGrayscale(string imagePath, int width, int height)
        {
            BitmapImage bitmap = new BitmapImage();
            bitmap.BeginInit();
            bitmap.CacheOption = BitmapCacheOption.OnLoad;
            bitmap.UriSource = new Uri(imagePath, UriKind.Absolute);
            bitmap.DecodePixelWidth = width;
            bitmap.DecodePixelHeight = height;
            bitmap.EndInit();
            bitmap.Freeze();
            FormatConvertedBitmap grayBitmap = new FormatConvertedBitmap(
                bitmap, 
                System.Windows.Media.PixelFormats.Gray8, 
                null, 
                0
            );

            byte[] pixels = new byte[width * height];
            grayBitmap.CopyPixels(pixels, width, 0);
            byte[,] result = new byte[height, width];
            for (int i = 0; i < height; i++)
            {
                for (int j = 0; j < width; j++)
                {
                    result[i, j] = pixels[i * width + j];
                }
            }

            return result;
        }
        private double[,] ComputeDCT(double[,] input)
        {
            int N = input.GetLength(0);
            int M = input.GetLength(1);
            double[,] output = new double[N, M];

            for (int u = 0; u < N; u++)
            {
                for (int v = 0; v < M; v++)
                {
                    double sum = 0.0;
                    for (int i = 0; i < N; i++)
                    {
                        for (int j = 0; j < M; j++)
                        {
                            sum += input[i, j] *
                                   Math.Cos((2 * i + 1) * u * Math.PI / (2 * N)) *
                                   Math.Cos((2 * j + 1) * v * Math.PI / (2 * M));
                        }
                    }

                    double cu = (u == 0) ? 1.0 / Math.Sqrt(N) : Math.Sqrt(2.0 / N);
                    double cv = (v == 0) ? 1.0 / Math.Sqrt(M) : Math.Sqrt(2.0 / M);
                    output[u, v] = cu * cv * sum;
                }
            }

            return output;
        }
        private string BitsToHex(bool[] bits)
        {
            StringBuilder binary = new StringBuilder();
            foreach (bool bit in bits)
            {
                binary.Append(bit ? '1' : '0');
            }
            while (binary.Length % 4 != 0)
            {
                binary.Append('0');
            }
            StringBuilder hex = new StringBuilder();
            for (int i = 0; i < binary.Length; i += 4)
            {
                string chunk = binary.ToString(i, 4);
                int value = Convert.ToInt32(chunk, 2);
                hex.Append(value.ToString("x"));
            }

            return hex.ToString();
        }
        /// <param name="hash1">First hash.</param>
        /// <param name="hash2">Second hash.</param>
        public static int HammingDistance(string hash1, string hash2)
        {
            if (hash1.Length != hash2.Length)
            {
                throw new ArgumentException("Hashes must be the same length");
            }
            long value1 = Convert.ToInt64(hash1, 16);
            long value2 = Convert.ToInt64(hash2, 16);
            long xor = value1 ^ value2;
            int distance = 0;
            while (xor != 0)
            {
                distance += (int)(xor & 1);
                xor >>= 1;
            }

            return distance;
        }

        /// <param name="hash1">First hash</param>
        /// <param name="hash2">Second hash</param>
        /// <param name="maxBits">Maximum number of bits</param>
        public static double CompareHashes(string hash1, string hash2, int maxBits = 64)
        {
            int distance = HammingDistance(hash1, hash2);
            return 1.0 - ((double)distance / maxBits);
        }
        /// <param name="hash1">First hash.</param>
        /// <param name="hash2">Second hash.</param>
        /// <param name="threshold">Similarity threshold</param>
        /// <param name="maxBits">Maximum number of bits</param>
        public static bool AreSimilar(string hash1, string hash2, double threshold = 0.9, int maxBits = 64)
        {
            double similarity = CompareHashes(hash1, hash2, maxBits);
            return similarity >= threshold;
        }
        public class HashData
        {
            public string ImagePath { get; set; } = string.Empty;
            public string Method { get; set; } = string.Empty;
            public string Hash { get; set; } = string.Empty;
            public int HashSize { get; set; }
        }
        /// <param name="imagePath">Path to the img</param>
        /// <param name="outputPath">Path to save the hash</param>
        /// <param name="method">Hash method</param>
        public void SaveHashToFile(string imagePath, string outputPath, HashMethod method = HashMethod.PerceptualHash)
        {
            string hash = ComputeHash(imagePath, method);

            var hashData = new HashData
            {
                ImagePath = Path.GetFullPath(imagePath),
                Method = method.ToString(),
                Hash = hash,
                HashSize = hashSize
            };

            string json = JsonSerializer.Serialize(hashData, new JsonSerializerOptions { WriteIndented = true });
            File.WriteAllText(outputPath, json);
        }
        /// <param name="hashFile">Path to the hash JSON file.</param>
        /// <returns>Hash data.</returns>
        public static HashData LoadHashFromFile(string hashFile)
        {
            string json = File.ReadAllText(hashFile);
            return JsonSerializer.Deserialize<HashData>(json) ?? throw new InvalidOperationException("Failed to deserialize hash data");
        }
        /// <param name="imagePath">Path to the img</param>
        /// <param name="hashFile">Path to the hash</param>
        /// <param name="threshold">Similarity threshold</param>
        /// <returns>Tuple of (isMatch, similarity).</returns>
        public (bool isMatch, double similarity) VerifyImage(string imagePath, string hashFile, double threshold = 0.9)
        {
            HashData hashData = LoadHashFromFile(hashFile);
            if (!Enum.TryParse<HashMethod>(hashData.Method, out HashMethod method))
            {
                method = HashMethod.PerceptualHash;
            }
            string currentHash = ComputeHash(imagePath, method);
            int maxBits = hashData.HashSize * hashData.HashSize;
            double similarity = CompareHashes(hashData.Hash, currentHash, maxBits);
            bool isMatch = similarity >= threshold;

            return (isMatch, similarity);
        }
    }
}
